### PR TITLE
Increase Resiliency Of Adaptive Scheduling For Occasional Slowdowns

### LINKED ISSUE #7074: Non-blocking Code Being Shifted To Autoblocking Pool, Degrading CPU Performance By Running Too Many Tasks In Parallel
I'm working on migrating a large production workload from ZIO 1.x to 2.x, and am encountering issues with the autoblocking detection. The detection seems to be picking up on things that are not blocking and shifting them to the blocking threadpool, which in turn is resulting in the app paralleling at a higher factor than it used to, overloading the CPU, and slowing the entire application.

My reading of the autoblocking implementation suggests that as it exists today, the definition of "blocking" is "took more than 100ms to execute", which is certainly not a compatible definition with our app. Something that is taking over 100ms because of its CPU usage is exactly the kind of thing that we want to keep on a threadpool that is bounded by number of processors. I also have some concerns around whether a long GC pause might cause something to be incorrectly flagged, though there may be reasons I'm missing that make that one a non issue. Finally, the mechanism for determining whether or not 100ms have elapsed [looks to be based on `System.currentTimeMillis()`](https://github.com/zio/zio/blob/series/2.x/core/jvm/src/main/scala/zio/internal/ZScheduler.scala#L208), which in our environment might fluctuate as the system clock updates, again potentially resulting in something being incorrectly flagged as "blocking".

Admittedly, I may be totally misunderstanding the code, so I don't want to get too deep into speculation around what the issue might be. But regardless, this has been my observation, and I've written an example app to show it. All this app really does is 1. runs a deep fibonacci computation with tail recursion, designed to take over 100ms to run (you may have to modify the example's depth based on your hardware), 2. tries to run 10,000 of them in parallel, 3. they get incorrectly detected as blocking and begin to submit to a blocking threadpool, 4. CPU goes to 100% and the app eventually crashes.

Running the same app on the latest ZIO 1.x version just takes a long time and spins the CPU up to *near* 100% (as expected) but does not crash.

I'm hoping that we can at least get a configuration option to disable autoblocking detection (which is probably a good option to have whether this can be fixed or not). 

```scala
package zio.scheduler.test

import zio.{Scope, ZIO, ZIOAppArgs}

import scala.annotation.tailrec

object BlockingSchedulerTest extends zio.ZIOAppDefault {

  override def run: ZIO[ZIOAppArgs with Scope, Any, Any] = {
    ZIO.foreachParDiscard(0 until 10000)(i => runFibThreadInfo(s"$i"))
  }

  def runFibThreadInfo(id: String) = runThreadInfo(id, runFibTimed(1000000000))

  def runThreadInfo[R, E, A](id: String, task: => ZIO[R, E, A]) = {
    for {
      beforeThread <- ZIO.succeed(Thread.currentThread().getName)
      _ <- task
      afterThread <- ZIO.succeed(Thread.currentThread().getName)
      _ <- zio.Console.printLine(s"$id started on thread $beforeThread and ended on thread $afterThread")
    } yield ()
  }

  def runFibTimed(n: Long) = {
    for {
      start <- zio.Clock.nanoTime
      _ <- ZIO.succeed(fib(n))
      end <- zio.Clock.nanoTime
      time = scala.concurrent.duration.Duration.fromNanos(end - start)
      _ <- zio.Console.printLine(s"took ${time.toMillis} ms")
    } yield ()
  }

  def fib(n: Long): Long = {
    @tailrec
    def fib_tail(n: Long, a: Long, b: Long): Long = n match {
      case 0 => a
      case _ => fib_tail(n - 1, b, a + b)
    }

    fib_tail(n, 0, 1)
  }
}
```

### PR DESCRIPTION
**Pull request by adamgfraser**
Resolves #7074.

Occasional slowdowns can cause a fiber fork location to be incorrectly identified as blocking. This PR modifies the heuristic to identify a fiber fork location as blocking to require a certain frequency of work from a location to be identified as blocking before identifying that location as blocking.

@dccorona Do you want to try this branch?

### CODE REVIEW

**Comment by dccorona** [comment]
Unfortunately, our internal processes for bringing in open-source code make it very difficult for us to bring in changes that aren't present in maven central. I'll look into what testing this patch might look like.

**Comment by adamgfraser** [comment]
Okay, I tested with this:

```scala
object Example extends ZIOAppDefault {

  def worker(activeProcesses: Ref[Int]): ZIO[Any, Nothing, Unit] =
    activeProcesses.updateAndGet(_ + 1) *>
      Random.nextIntBounded(100).flatMap { n =>
        if (n == 0) ZIO.succeed(Thread.sleep(1000))
        else ZIO.unit
      } *>
      activeProcesses.update(_ - 1)

  val run =
    for {
      activeProcesses <- Ref.make(0)
      _               <- ZIO.collectAllParDiscard(Chunk.fill(100000)(worker(activeProcesses)))
    } yield ()
}
```

**Comment by aartigao** [comment]
We can try the fix, but it's weekend and that means we'll have to wait until Monday.  I'll keep you posted☝🏼

File: core/jvm/src/main/scala/zio/internal/ZScheduler.scala [hunk 1]
```diff
@@ -205,8 +208,9 @@ private final class ZScheduler extends Executor {
   private[this] def makeSupervisor(): ZScheduler.Supervisor =
     new ZScheduler.Supervisor {
       override def run(): Unit = {
-        var currentTime      = java.lang.System.currentTimeMillis()
-        val previousOpCounts = Array.fill(poolSize)(-1L)
+        var currentTime         = java.lang.System.currentTimeMillis()
```
Comment (aartigao) [inline core/jvm/src/main/scala/zio/internal/ZScheduler.scala:211]: Out of curiosity, why not `nanoTime`?. I've always heard that measuring with `currentTimeMillis` is known to cause [issues](https://www.javaadvent.com/2019/12/measuring-time-from-java-to-kernel-and-back.html).

**Comment by adamgfraser** [comment]
@aartigao That would be great!

**Comment by aartigao** [comment]
Tomorrow I'll give you more detailed information, but for the moment results are better/_acceptable_.

Still, the numbers are worse than RC5, but it's a tiny difference.

**Comment by adamgfraser** [comment]
@aartigao That's great!

**Comment by aartigao** [comment]
Hi again! 👋🏽  Following graphic is the 90th percentile comparison between RC5 (yellow) and this PR's fix (green). Our peak hours are from 18:30h to 23h.

image

As said, the difference is not that big as it used to be without the fix, but still you can see there has been a slowdown on peak hours 👎🏽

**Comment by adamgfraser** [comment]
@aartigao Thanks for reporting! There were a lot of other changes in between `RC5` and the final version of ZIO, in particular an entirely new runtime system, so I think we probably need to try to disentangle the impacts of those before making further changes. I created a branch without autoblocking [here](https://github.com/adamgfraser/zio/tree/noautoblocking) if you want to try that. Thanks again for doing this and feel free to reach out to me directly.

**Comment by aartigao** [comment]
>I created a branch without autoblocking [here](https://github.com/adamgfraser/zio/tree/noautoblocking) if you want to try that

Sure! We'll try it and let you know 🙂

**Review by jdegoes — APPROVED** [review]
We should investigate the possibility of feeding runtime flags into the scheduler so we have some scoped control over features like auto-blocking.

**Pull request merged**